### ISSUE TITLE
[Bug]: Mixed prompt embedding masks are omitted from prefix cache keys, causing incorrect outputs

### ISSUE THREAD

**Issue by ai-yang** [vllm-project/vllm#56655]
## Your current environment

- Tested source: `main` at `a2bc2ffb2c8bf46759c4bbeb88e2f4dc32968762`.
- Python 3.12.14; PyTorch 2.13.0+cu132; Transformers 5.16.1.
- NVIDIA GeForce RTX 3090; driver 580.173.02; one GPU; FP16.
- vLLM editable source with the official **cu130 precompiled extension for the exact source commit**. This combination was exercised successfully in the GPU tests below.
- Model: `Qwen/Qwen3-0.6B`, revision `c1899de289a04d12100db370d81485cdf75e47ca`, using its real embedding weights.
- V2 Model Runner, FLASH_ATTN, eager execution, block size 16, greedy sampling, seed 0. `VLLM_WORKER_MULTIPROC_METHOD=spawn`, `OMP_NUM_THREADS=1`, `VLLM_USE_FLASHINFER_SAMPLER=0`.

The original full before/after experiment used driver 580.173.02. The additional standalone-script runs on 2026-09-13 used driver 580.178.04 and produced the same token/cache results. The host runs Ubuntu 22.04.4 LTS, glibc 2.35; the current `collect_env` reports CUDA toolkit 13.2.78. These are separately recorded runs rather than one unchanged host snapshot.

The affected unpatched function is also unchanged at `e52be1a62d3879b1202f4f355d3c3472b560c6f2`, checked on 2026-09-13. The numerical results below are from the explicitly identified tested revision, not a claim of rerunning the entire later `main`.

## Describe the bug

With `enable_prompt_embeds=True` and `enable_prefix_caching=True`, valid inputs with the same token IDs and supplied embedding tensor but different `prompt_is_token_ids` masks can reuse KV computed for a different effective input. The request succeeds, but its deterministic output depends on an earlier request's cache contents.

`prompt_is_token_ids[i]` selects either the model's learned embedding for the token ID (`True`) or the supplied embedding row (`False`). The prefix-cache key hashes the normalized token IDs and supplied tensor, but omits that selection mask. Embedding-backed positions are normalized to token ID 0, which is also a valid vocabulary ID.

In the reproducer, B alone generates `[9, 0] * 8`. After caching A, the **identical B** incorrectly hits 32 cached tokens and generates `[753] * 16`. Repeating A itself correctly hits 32 tokens.

### Scope

This is a functional correctness issue in the combination of mixed prompt embeddings and local prefix caching. The reproducer uses the public offline input interface, without external KV connectors, LoRA, concurrency, or stochastic sampling. It does not establish a failure in ordinary text-only requests or a production occurrence rate.

The standard chat renderer zero-fills supplied-tensor rows at text positions, which narrows the conditions for two rendered requests to have identical tensor bytes and different masks. A complete HTTP Chat Completions reproduction is not claimed here. Temporarily disabling prefix caching avoids this erroneous reuse.

## Reproduction

In a working vLLM development environment, obtain the model snapshot and save the following code as `issue_cache_mask.py`:

```bash
hf download Qwen/Qwen3-0.6B \
  --revision c1899de289a04d12100db370d81485cdf75e47ca \
  --local-dir ./Qwen3-0.6B
CUDA_VISIBLE_DEVICES=0 .venv/bin/python issue_cache_mask.py ./Qwen3-0.6B
```

The script asserts correct behavior. On the unpatched revision it fails at the `BUG: B's output changed...` assertion in the first trial. With the fix it completes all three trials. The configured memory fraction requires approximately 7.1 GiB of free memory on the tested 24 GiB GPU.

```python
"""Run with the audited environment: .venv/bin/python issue_cache_mask.py MODEL_DIR.

The unpatched revision fails the final cache/output assertions; the fix passes.
"""

import json
import os
import sys
from pathlib import Path

os.environ["VLLM_WORKER_MULTIPROC_METHOD"] = "spawn"
os.environ["OMP_NUM_THREADS"] = "1"
os.environ["VLLM_USE_FLASHINFER_SAMPLER"] = "0"

import torch
from safetensors import safe_open

from vllm import LLM, SamplingParams


def main():
    model_path = sys.argv[1]  # Local Qwen/Qwen3-0.6B snapshot.
    with safe_open(
        str(Path(model_path) / "model.safetensors"), framework="pt", device="cpu"
    ) as weights:
        row = weights.get_slice("model.embed_tokens.weight")[198:199]
        supplied = row.to(torch.float16).repeat(33, 1)

    llm = LLM(
        model=model_path,
        dtype="half",
        enable_prompt_embeds=True,
        enable_prefix_caching=True,
        enforce_eager=True,
        block_size=16,
        max_model_len=256,
        max_num_seqs=1,
        gpu_memory_utilization=0.3,
        seed=0,
    )
    params = SamplingParams(temperature=0, max_tokens=16, logprobs=20, seed=0)

    def generate(case, use_tokens):
        # A uses row 198 at the first 32 positions; B uses token ID 0.
        # Both have the same token IDs and supplied tensor. Only the mask differs.
        prompt = {
            "prompt_token_ids": [0] * 33,
            "prompt_embeds": supplied.clone(),
            "prompt_is_token_ids": [use_tokens] * 32 + [True],
        }
        result = llm.generate([prompt], params, use_tqdm=False)[0]
        record = {
            "case": case,
            "cached_tokens": result.num_cached_tokens,
            "token_ids": list(result.outputs[0].token_ids),
        }
        print(json.dumps(record), flush=True)
        return record

    for trial in range(3):
        print(f"trial={trial}", flush=True)
        assert llm.reset_prefix_cache()
        cold_b = generate("B_cold", True)
        assert llm.reset_prefix_cache()
        cold_a = generate("A_cold", False)
        warm_a = generate("A_after_A", False)
        warm_b = generate("B_after_A", True)

        assert cold_a["cached_tokens"] == cold_b["cached_tokens"] == 0
        assert cold_a["token_ids"] != cold_b["token_ids"], "Control inputs must differ"
        assert warm_a["cached_tokens"] == 32
        assert warm_a["token_ids"] == cold_a["token_ids"]
        assert warm_b["token_ids"] == cold_b["token_ids"], (
            "BUG: B's output changed after caching A with a different routing mask"
        )
        assert warm_b["cached_tokens"] == 0, "B must not reuse A's two cache blocks"


if __name__ == "__main__":
    main()
```

Both requests supply `[0] * 33` and the same tensor containing 33 copies of the model's learned embedding row 198. A uses external embeddings for the first 32 positions and token ID 0 for the final position; B uses token ID 0 everywhere. Only the routing mask differs.

As an additional control in the full experiment, A's cold-cache output equals the output of token IDs `[198] * 32 + [0]`, and B's cold-cache output equals that of `[0] * 33`. This checks that the mixed-input construction has the intended model semantics.

## Actual and expected results

The full GPU experiment completed three independent trials before and three after the fix. The compact script above was also checked separately: it fails at the expected assertion without the fix and passes all three trials with it.

| Request | Unpatched: cached tokens / generated token IDs | Fixed: cached tokens / generated token IDs |
| --- | --- | --- |
| B, cold cache | 0 / `[9, 0] * 8` | 0 / `[9, 0] * 8` |
| A, cold cache | 0 / `[753] * 16` | 0 / `[753] * 16` |
| A after A | 32 / `[753] * 16` | 32 / `[753] * 16` |
| B after A | **32 / `[753] * 16`, incorrect** | **0 / `[9, 0] * 8`, matches B's cold output** |

Expected: KV computed from a different effective input must not be reused. Identical requests should retain normal prefix-cache hits. These deliberately controlled inputs demonstrate correctness, not general language-model quality or task accuracy.

## Root cause and proposed fix

[`Request`](https://github.com/vllm-project/vllm/blob/a2bc2ffb2c8bf46759c4bbeb88e2f4dc32968762/vllm/v1/request.py#L147) normalizes external-embedding positions to token ID 0. [`_gen_prompt_embeds_extra_hash_keys()`](https://github.com/vllm-project/vllm/blob/a2bc2ffb2c8bf46759c4bbeb88e2f4dc32968762/vllm/v1/core/kv_cache_utils.py#L559) includes only the tensor digest. Consequently, the two different effective inputs above have identical hash inputs.

Include the block's routing mask as a separate structured extra-key element when it contains token-ID positions. Preserve equivalence between a pure-embeddings input and an explicit all-False mask. Keeping the mask separate from the tensor bytes avoids ambiguous concatenation at a prompt/decode boundary.

## Related issues and PRs checked

Rechecked on 2026-09-13: all-state issues for `"prompt_is_token_ids" "cache"`, all-state PRs for `"prompt_is_token_ids" "hash"`, open PRs for `"prompt_embeds" "cache"`, and open PRs referencing `42119`.

- #42119 concerns **external LMCache** and different embedding tensor contents, with local prefix caching disabled in the report. This issue concerns local cache reuse with identical tensors and different routing masks.
- #51899 adds type tags to extra-key fields; its current patch does not include the routing mask.
- #53769 adds cache-key conformance tests; its current patch neither changes this production hash nor covers this mask distinction.
- #56639 fixes prompt-embedding inputs in decoder pooling; it does not change cache hashing.
- #44842 and #41564 concern input-position initialization and wake-up handling respectively, rather than the cache identity omission here.

No corresponding mask-hashing fix was found in these checks.

## AI assistance

AI assistance was used for investigation, reproduction, the proposed fix, and preparation of this report. The test results above are from executed tests; the limits of those tests are stated explicitly.

**Comment by ZC502** [comment]
@ai-yang This is a nice example of a runtime-state axis that is easy to lose in ordinary parity testing: the logical B request is unchanged, but `B_cold` and `B_after_A` differ only by prefix-cache history.

I’ve been extending a small measurement-only parity format for vLLM, and this looks like a useful **cache-state parity** case: preserve request identity, cache precondition, `num_cached_tokens`, generated token IDs, and sampled logprobs as separate evidence rather than treating the warm-cache result as just another arm.

Your reproducer is already controlled enough that I would not try to classify root cause from the traces — the hash-key investigation is much stronger evidence for that. But if you’re willing to expose the small JSON output from one before/after run, I’d be interested in adapting it as another external-probe fixture, similar to how we’re handling batch-composition traces in #56370.

**Comment by yigitcan-ozturk** [comment]
Hi, I’d like to take this correctness bug. The failure mode and proposed invariant look well-bounded: two mixed prompt-embedding requests can have identical token/tensor hash inputs while differing in `prompt_is_token_ids`, so local prefix caching may reuse KV for a different effective input.

I’ll keep the scope focused on cache identity. I plan to trace `_gen_prompt_embeds_extra_hash_keys()` and its block slicing/callers, add CPU-level hash/key regression coverage for masks that differ within a cache block (including all-False equivalence where intended), then make the smallest structured-key change that includes the routing mask without changing ordinary text-only cache keys. If the existing test harness supports the mixed-input path without a GPU, I’ll add that coverage too; otherwise I’ll keep the executed GPU reproduction from this report as the end-to-end reference rather than claiming hardware validation I cannot run.

I checked the issue discussion and current open-PR search and did not find an active implementation for #56655. I’ll proceed unless there is work in progress elsewhere.
